src/carma-platform/guidance_scripts.py

```diff
@@ -1283,6 +1283,352 @@ def get_planner_trajectory_intervals(
+    ax2 = plt.subplot(2, 1, 2, sharex=ax1)
+    speed_error = long_velocities - speed_limits
+    ax2.plot(timestamps, speed_error, 'r-', linewidth=1.5)
+    ax2.axhline(y=0, color='k', linestyle='-', alpha=0.3)
+
+
+    # Highlight tolerance bounds - only add to legend once
+    tolerance_added_to_legend = False
+    for event in speed_limit_changes:
+        event_time, old_limit, new_limit = event
+        # Draw tolerance bands after the speed change
+        upper_tolerance = new_limit * speed_tolerance_pct
+        lower_tolerance = -new_limit * speed_tolerance_pct
+
+        if not tolerance_added_to_legend:
+            # First time, add to legend
+            ax2.axhline(y=upper_tolerance, color='g', linestyle='--', alpha=0.5,
+                      label=f'Tolerance (Â±{speed_tolerance_pct*100:.0f}%)')
+            ax2.axhline(y=lower_tolerance, color='g', linestyle='--', alpha=0.5)
+            tolerance_added_to_legend = True
+        else:
+            # Subsequent times, don't add to legend
+            ax2.axhline(y=upper_tolerance, color='g', linestyle='--', alpha=0.5)
+            ax2.axhline(y=lower_tolerance, color='g', linestyle='--', alpha=0.5)
+
+    ax2.set_title("Speed Error (Vehicle Speed - Speed Limit)")
+    ax2.set_xlabel(TIME_SECONDS_LABEL_STRING)
+    ax2.set_ylabel("Error (m/s)")
+    ax2.grid(True, alpha=0.3)
+    ax2.legend()
+
+    plt.tight_layout()
+
+    # Save results if directories are provided
+    if save_stats_dir:
+        stats_dir = Path(save_stats_dir)
+        stats_dir.mkdir(parents=True, exist_ok=True)
+        stats_path = stats_dir / "speed_change_analysis.json"
+        with open(stats_path, "w") as f:
+            json.dump({
+                "pass_results": pass_results,
+                "statistics": statistics
+            }, f, indent=2)
+        print(f"Stats saved to: {stats_path}")
+
+    if save_data_dir:
+        data_dir = Path(save_data_dir)
+        data_dir.mkdir(parents=True, exist_ok=True)
+        data_path = data_dir / "speed_change_analysis_data.npz"
+        np.savez(
+            data_path,
+            timestamps=timestamps,
+            velocities=long_velocities,
+            speed_limits=speed_limits,
+            cmd_velocities=cmd_velocities,
+            speed_limit_changes=speed_limit_changes,
+            response_times=response_times,
+            steady_state_periods=steady_state_periods
+        )
+        print(f"Data saved to: {data_path}")
+
+    if save_plot_dir:
+        plot_dir = Path(save_plot_dir)
+        plot_dir.mkdir(parents=True, exist_ok=True)
+        plot_path = plot_dir / "speed_change_analysis.png"
+        plt.savefig(plot_path, dpi=300)
+        print(f"Plot saved to: {plot_path}")
+    else:
+        plt.show()
+
+    return (pass_results["overall"], statistics, fig, speed_limit_changes, response_times)
+
+
+def detect_speed_limit_changes(timestamps, speed_limits, min_change=0.5):
```

Comment:
I haven't taken a full look at how we're using this file, but if we add a function to this file, do we need to document it in guidance_scripts.md? If it's just a helper function, can we keep it in guidance_scripts.py without documenting it, or should we consider moving it to utils.py? I am okay with keeping the functions here or moving them to utils.py, but wanted to get your thoughts (question applies for `analyze_speed_responses()` as well)

Reply:
In terms of documenting, we should only document if the helper function can be used somewhere else like `get_engage_time` or main functions like `run_some_analysis`. Utils.py in my mind had carma's stack agnostic scripts/tools that are applicable to all scripts like `calculate_error_statistics` (or maybe plotting in the future). So it seemed appopriate to keep the guidance helper functions in the guidance scripts as they are only related to guidance related functionalities. If the file is getting too big, maybe we could create guidance_helper.py or something like that.